### ISSUE TITLE
Setting a `color.palette` in a child theme wipes out parent colors

### ISSUE THREAD

**Issue by ryelle** [WordPress/gutenberg#40557]
### Description

When the parent and child theme `theme.json` files are merged, [the array values are not merged, but rather overwritten](https://github.com/WordPress/gutenberg/blob/ace064391f2ce2d96c30a0c8761a5d2a985b9c5b/lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php#L385). So if you have a parent theme with a set of 10 colors in `color.palette`, and a child theme that adds a single color in `color.palette`, those 10 parent colors are no longer set. 

This is apparently by design, as the long comment linked says, but I don't think this should be the case. As mentioned, this will happen to `color.palette, color.gradients, color.duotone, typography.fontSizes, or typography.fontFamilies.` I think it would be common for a parent theme to set up colors and typography, and let the child theme extend those "base" settings. Maybe merging by slug?

### Step-by-step reproduction instructions

1. Create a parent theme and set some default colors
2. Activate this theme, see all your colors available in the editor
3. Create a child theme, add a theme.json file, and set just one or two colors
4. Activate the child theme, now see only the child theme colors

Or just add/remove the `color.palette` property from your child `theme.json`.

### Screenshots, screen recording, code snippet

This issue also means that existing parent theme templates that use parent colors suddenly don't have those colors anymore.

| Without `color.palette`, inheriting colors from parent | With a `color.palette` |
|-----|-----|
| ![](https://user-images.githubusercontent.com/541093/164768603-a83a4be6-aea5-42d8-a308-06c4cd9d4df6.png) | ![](https://user-images.githubusercontent.com/541093/164768600-c985b838-f840-4915-a9c5-6e3cb85af70d.png) |

### Environment info

- WP 6.0-beta2
- Gutenberg 13.0

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by carolinan** [comment]
It is the same with the style variations, if I add any color value, it overrides the theme.json and if I want "eveything+1" I have to add "everything+1"

I am not sure what the best solution is though, because both scenarios are valid.

**Comment by scruffian** [comment]
IMO the two arrays should be merged.

**Comment by MaggieCabrera** [comment]
> IMO the two arrays should be merged.

I'm not so sure, if you want a child theme to override all the parent's colors, there's no way for you to do that. If you want both, you still can do that even if you have to copy manually the parent's colors. What about having a new option to select if we want to merge or override?

**Comment by ryelle** [comment]
> If you want both, you still can do that even if you have to copy manually the parent's colors.

If the parent theme tweaks the colors, the child theme would also need to be updated. The benefit of using a child theme (IMO, of course) is that you can just magically inherit these kinds of settings.

I think the parent theme colors always need to be defined, since they might be used in parent templates (see the screenshots in the description).

What would be the drawback of merging? I would expect it to merge by slug, so if the child theme wants to manage all the colors, it could redefine them all in its `theme.json`.

**Comment by MaggieCabrera** [comment]
> > If you want both, you still can do that even if you have to copy manually the parent's colors.
> 
> If the parent theme tweaks the colors, the child theme would also need to be updated. The benefit of using a child theme (IMO, of course) is that you can just magically inherit these kinds of settings.
> 
> I think the parent theme colors always need to be defined, since they might be used in parent templates (see the screenshots in the description).
> 
> What would be the drawback of merging? I would expect it to merge by slug, so if the child theme wants to manage all the colors, it could redefine them all in its `theme.json`.

yeah, I see what you mean, if the slug is the same we keep the child theme value, that makes sense. I was thinking of having double the amount of colors.

**Comment by pbking** [comment]
There are definitely scenarios where we would want FEWER colors.  Blockbase, for example, leverages 5 palette colors (primary, secondary, foreground, background tertiary) but many of its child themes have fewer (Quadrat, for instance, only has primary, tertiary and background).

I think it's important that Blockbase provide the OPTION to style elements with the control that 5 color options gives, so I wouldn't want to limit that.  But it's also right that Quadrat only needs to provide the user with three colors.  To override two of the Blockbase colors with repeated values wouldn't seem right.

I keep flip-flopping.  I like the way it is now because it's simple to control what goes out to the user.  Merging the two collections makes sense too.  I'm not sure how to account for both scenarios.

**Comment by gturpin-dev** [comment]
Hi,
Obviously, it can't be by default the palette defined in parent theme be merged with the one defined in the child because in this case there is no way to remove on color that we don't want from the parent.

The same issue happens with the font sizes and probably other array variables.

I suggest something like : every array variables in theme.json can have an additionnal setting like "mergeWithParent" or something like that.

False by default and it will redefine the parent palette. But if true we merge the parent palette and then adding only the one we want.

All case are handled, and of course if you want to remove only one color, you should copy the entire parent palette manually, but I don't think we have another easy options.

**Comment by t-hamano** [comment]
I would like to mention that by using [wp_theme_json_data_theme](https://developer.wordpress.org/reference/hooks/wp_theme_json_data_theme/) hook, you can both overwrite and merge the parent theme's palette. With this in mind, I think we need to consider whether overwriting or merging makes more sense.

### Override

```php
<?php
function custom_wp_theme_json_theme( $theme_json ) {
	$new_data = array(
		'version'  => 2,
		'settings' => array(
			'color'  => array(
				'palette'  => array(
					array(
						'color' => '#ff0000',
						'name'  => 'Child Red',
						'slug'  => 'red',
					)
				),
			),
		),
	);
	return $theme_json->update_with( $new_data );
}
add_filter( 'wp_theme_json_data_theme', 'custom_wp_theme_json_theme' );
```

### Merge

```php
<?php
function custom_wp_theme_json_theme( $theme_json ) {
	$parent_theme_json_data = $theme_json->get_data();
	$child_palette          = array(
		array(
			'color' => '#ff0000',
			'name'  => 'Child Red',
			'slug'  => 'red',
		),
	);
	if ( isset( $parent_theme_json_data['settings']['color']['palette']['theme'] ) ) {
		$child_palette = array_merge(
			$parent_theme_json_data['settings']['color']['palette']['theme'],
			$child_palette,
		);
	}

	$new_data = array(
		'version'  => 2,
		'settings' => array(
			'color'  => array(
				'palette'  => $child_palette,
			),
		),
	);
	return $theme_json->update_with( $new_data );
}
add_filter( 'wp_theme_json_data_theme', 'custom_wp_theme_json_theme' );
```

**Comment by Simeon-development-envir** [comment]
@ryelle https://phps.space/questions/customizing-a-wordpress-parent-theme-using-a-child-theme

**Comment by youknowriad** [comment]
Closing as working as designed, with the workaround documented in-thread: array values in theme.json intentionally replace rather than merge , and the wp_theme_json_data_theme filter lets a child theme merge the parent palette explicitly when that's the intent. I'd rather avoid additional complexity on this. Thanks all!
